packages/eslint-plugin/src/rules/naming-convention.ts

```diff
@@ -0,0 +1,1212 @@
+function parseOptions(context: Context): ParsedOptions {
+  const normalizedOptions = context.options.map(opt => normalizeOption(opt));
+  const parsedOptions = util.getEnumNames(Selectors).reduce((acc, k) => {
+    acc[k] = createValidator(k, context, normalizedOptions);
+    return acc;
+  }, {} as ParsedOptions);
+
+  return parsedOptions;
+}
+function createValidator(
+  type: SelectorsString,
+  context: Context,
+  allConfigs: NormalizedSelector[],
+): (node: TSESTree.Identifier | TSESTree.Literal) => void {
+  // make sure the "highest priority" configs are checked first
+  const selectorType = Selectors[type];
+  const configs = allConfigs
+    // gather all of the applicable selectors
+    .filter(
+      c =>
+        (c.selector & selectorType) !== 0 ||
+        c.selector === MetaSelectors.default,
+    )
+    .sort((a, b) => {
+      if (a.selector === b.selector) {
+        // in the event of the same selector, order by modifier weight
+        // sort ascending - the type modifiers are "more important"
+        return b.modifierWeight - a.modifierWeight;
+      }
+
+      // sort descending - the meta selectors are "least important"
+      return a.selector - b.selector;
+    });
+
+  return (
+    node: TSESTree.Identifier | TSESTree.Literal,
+    modifiers: Set<Modifiers> = new Set<Modifiers>(),
+  ): void => {
+    const originalName =
+      node.type === AST_NODE_TYPES.Identifier ? node.name : `${node.value}`;
+
+    // return will break the loop and stop checking configs
+    // it is only used when the name is known to have failed or succeeded a config.
+    for (const config of configs) {
+      if (config.filter?.test(originalName)) {
+        // name does not match the filter
+        continue;
+      }
+
+      if (config.modifiers?.some(modifier => !modifiers.has(modifier))) {
+        // does not have the required modifiers
+        continue;
+      }
+
+      if (!isCorrectType(node, config, context)) {
+        // is not the correct type
+        continue;
+      }
+
+      let name: string | null = originalName;
+
+      name = validateUnderscore('leading', config, name, node, originalName);
+      if (name === null) {
+        // fail
+        return;
+      }
+
+      name = validateUnderscore('trailing', config, name, node, originalName);
+      if (name === null) {
+        // fail
+        return;
+      }
+
+      name = validateAffix('prefix', config, name, node, originalName);
+      if (name === null) {
+        // fail
+        return;
+      }
+
+      name = validateAffix('suffix', config, name, node, originalName);
+      if (name === null) {
+        // fail
+        return;
+      }
+
+      if (!validatePredefinedFormat(config, name, node, originalName)) {
+        return;
+      }
+
+      // it's valid for this config, so we don't need to check any more configs
+      return;
+    }
+  };
+
+  // centralises the logic for formatting the report data
```

Comment:
```suggestion
  // centralize the logic for formatting the report data
```

Reply:
running it locally I get no errors...

Looking into it...
```bash
$ yarn check:spelling -v | grep eslint-plugin/src/rules
CSpell: Files checked: 104, Issues found: 0 in 0 files
```

```bash
$ yarn check:spelling -v | grep eslint-plugin/docs
Checking: /Users/bradzacher/github/typescript-eslint/packages/eslint-plugin/docs/rules/adjacent-overload-signatures.md, File type: markdown, Language: en-US ... Issues: 0 0.007S
Checking: /Users/bradzacher/github/typescript-eslint/packages/eslint-plugin/docs/rules/array-type.md, File type: markdown, Language: en-US ... Issues: 0 0.018S
Checking: /Users/bradzacher/github/typescript-eslint/packages/eslint-plugin/docs/rules/await-thenable.md, File type: markdown, Language: en-US ... Issues: 0 0.012S
Checking: /Users/bradzacher/github/typescript-eslint/packages/eslint-plugin/docs/rules/ban-ts-ignore.md, File type: markdown, Language: en-US ... Issues: 0 0.011S
Checking: /Users/bradzacher/github/typescript-eslint/packages/eslint-plugin/docs/rules/ban-types.md, File type: markdown, Language: en-US ... Issues: 0 0.011S

# ... snip ...

CSpell: Files checked: 104, Issues found: 0 in 0 files
```

Reply:
Ah, it's a mac thing.
you used the CLI command `cspell --config=.cspell.json **/*.{md,ts,js}`
macos auto-expands the glob, which doesn't expand correctly. adding quotes around the command makes it work correctly.

broken:
```bash
bradzacher-mbp:typescript-eslint bradzacher$ yarn check:spelling
yarn run v1.19.1
$ cspell --config=.cspell.json -v **/*.{md,ts,js}

cspell;
Date: Mon, 30 Dec 2019 05:35:21 GMT
Options:
    verbose:   Yes
    config:    .cspell.json
    exclude:   node_modules/**
    files:     **/*.md,tools/generate-contributors.ts,**/*.js
    wordsOnly: No
    unique:    No

Config Files Found:
    .cspell.json
```

working:
```bash
bradzacher-mbp:typescript-eslint bradzacher$ yarn check:spelling
yarn run v1.19.1
$ cspell --config=.cspell.json -v "**/*.{md,ts,js}"

cspell;
Date: Mon, 30 Dec 2019 05:35:49 GMT
Options:
    verbose:   Yes
    config:    .cspell.json
    exclude:   node_modules/**
    files:     **/*.{md,ts,js}
    wordsOnly: No
    unique:    No

Config Files Found:
    .cspell.json
```

Reply:
Ah, it's a mac thing.
you used the CLI command `cspell --config=.cspell.json **/*.{md,ts,js}`
macos auto-expands the glob, which doesn't expand correctly. adding quotes around the command makes it work correctly.

broken:
```bash
bradzacher-mbp:typescript-eslint bradzacher$ yarn check:spelling
yarn run v1.19.1
$ cspell --config=.cspell.json -v **/*.{md,ts,js}

cspell;
Date: Mon, 30 Dec 2019 05:35:21 GMT
Options:
    verbose:   Yes
    config:    .cspell.json
    exclude:   node_modules/**
    files:     **/*.md,tools/generate-contributors.ts,**/*.js
    wordsOnly: No
    unique:    No

Config Files Found:
    .cspell.json
```

working:
```bash
bradzacher-mbp:typescript-eslint bradzacher$ yarn check:spelling
yarn run v1.19.1
$ cspell --config=.cspell.json -v "**/*.{md,ts,js}"

cspell;
Date: Mon, 30 Dec 2019 05:35:49 GMT
Options:
    verbose:   Yes
    config:    .cspell.json
    exclude:   node_modules/**
    files:     **/*.{md,ts,js}
    wordsOnly: No
    unique:    No

Config Files Found:
    .cspell.json
```

Reply:
i'm not a mac user 😉 i tested it on windows + linux